src/hotspot/share/gc/g1/g1BarrierSet.cpp

```diff
@@ -107,8 +107,12 @@ void G1BarrierSet::invalidate(MemRegion mr) {
   }
   volatile CardValue* byte = _card_table->byte_for(mr.start());
   CardValue* last_byte = _card_table->byte_for(mr.last());
-  // skip initial young cards
-  for (; byte <= last_byte && *byte == G1CardTable::g1_young_card_val(); byte++);
+
+  // skip young cards
+  if (*byte == G1CardTable::g1_young_card_val()) {
+    guarantee(*last_byte == G1CardTable::g1_young_card_val(), "MemRegion should be either completely in young gen or not");
```

Comment:
I think a stronger assertion is that `mr` is inside a heap-region, sth along the line of `mr.size() < HeapRegionSize`.